### PR TITLE
feat(DCbreadcrumb): add capability to display location (and use it on Cable)

### PR DESCRIPTION
**Pull request by stonebuzz**
Add capability to display location from ```DCBreadcrumb```

(and use it from ```Cable``` object)

Before : 
image

After : 
image

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #number

### CODE REVIEW

**Comment by orthagh** [comment]
Did you test with very long location name ?
if not, maybe some text-overflow classes should be added

**Comment by stonebuzz** [comment]
> text-overflow

(without change for the moment)

image

**Comment by stonebuzz** [comment]
> Did you test with very long location name ? if not, maybe some text-overflow classes should be added

Maybe better with ```text-truncate```

image

**Comment by btry** [comment]
Hi

Interesting feature. After viewing the rendering, I got an idea. The breadcrumb consumes 2 lines very fast, without the need of long strings. Then why not show the locations in the 2nd line directly, horizontally aligned with the matching breadcrumb item ?

Something like that : 
```preformatted 
General > Datacenter > Server room > Rack
          Marseille >  Marseille >   Marseille

```

**Comment by cconard96** [comment]
Realistically, how often is a rack going to have a different location than its server room or a server room with a different location than its datacenter?

I think some optimization can be done with that. If a breadcrumb item has the same location as the one before it, there shouldn't be a need to display the location again. If a breadcrumb item in in a child location of the item before it, only the difference in the complete name needs shown. If a breadcrumb item has no location, the badge can be hidden too so it appears that the location is inherited.
If there is some weird case where the location of two breadcrumb items are not in the same hierarchy, then the full location can be shown I guess.

I'm not sure how complicated that would be to implement though.

**Comment by stonebuzz** [comment]
> Hi
> 
> Interesting feature. After viewing the rendering, I got an idea. The breadcrumb consumes 2 lines very fast, without the need of long strings. Then why not show the locations in the 2nd line directly, horizontally aligned with the matching breadcrumb item ?
> 
> Something like that :
> 
> ```
> General > Datacenter > Server room > Rack
>           Marseille >  Marseille >   Marseille
> ```
> 
> or
> 
> ```
> General > Datacenter > Server room > Rack
>           Marseille  > Marseille   > Marseille
> ```

new version 

image

**Comment by orthagh** [comment]
How the breadcrumb behaves when there is no location for an item ?

**Comment by stonebuzz** [comment]
> Realistically, how often is a rack going to have a different location than its server room or a server room with a different location than its datacenter?
> 
> I think some optimization can be done with that. If a breadcrumb item has the same location as the one before it, there shouldn't be a need to display the location again. If a breadcrumb item in in a child location of the item before it, only the difference in the complete name needs shown. If a breadcrumb item has no location, the badge can be hidden too so it appears that the location is inherited. If there is some weird case where the location of two breadcrumb items are not in the same hierarchy, then the full location can be shown I guess.
> 
> I'm not sure how complicated that would be to implement though.

@cconard96 

Some examples:

GLPI location tree

image

Datacenter -> ```A```
DCRoom -> ```A > A1```
Rack -> ```B > B1```

Breadcrumb

image



Datacenter -> ```A```
DCRoom -> ```A```
Rack -> ```A1```

Breadcrumb 

image

**Comment by stonebuzz** [comment]
> How the breadcrumb behaves when there is no location for an item ?

image

**Comment by stonebuzz** [comment]
> It may be good to prepend child locations with "> " to differentiate them from completely separate locations.

Like this

image

File: src/Features/DCBreadcrumb.php [hunk 1]
```diff
@@ -171,10 +262,32 @@ private function getItemEnclosurePosition($itemtype, $items_id)
...
 
+    /**
+     * get item location from Enclosure
+     *
+     * @param string  $itemtype Item type
+     * @param integer $items_id Item ID
+     *
+     * @return string
+     */
+    private function getItemLocation($itemtype, $items_id)
```
Comment (orthagh) [inline src/Features/DCBreadcrumb.php:277]: a suggestion here, this method seems not related to DCBreadcrumb class and doesn't require any `$this`.
I suggest to move it to Location class as private method.
Like `Location::getForItem(CommonDBTM $item)`  (avoid itemtype, items_id parameter when you the item initialy)

File: templates/layout/parts/dcbreadcrumbs.html.twig [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+         <div class="col-auto p-1">
+            >
+         </div>
+      {% else %}
+         <div class="col-auto pt-1 p-0">
+            {{ breadcrumb_item['position']|raw }}
+         </div>
+      {% endif %}
+   {% endfor %}
+</div>
```
Comment (orthagh) [inline templates/layout/parts/dcbreadcrumbs.html.twig:56]:
```suggestion
</div>

```

File: src/Features/DCBreadcrumb.php [hunk 3]
```diff
@@ -111,11 +142,71 @@ public function getDcBreadcrumb()
...
+        $breadcrumb = array_reverse($breadcrumb);
+
+        if ($with_location) {
+            $breadcrumb = $this->cleanLocationInforamtions($breadcrumb);
+        }
+
+        return $breadcrumb;
+    }
+
+    public function cleanLocationInforamtions(&$breadcrumb)
```
Comment (orthagh) [inline src/Features/DCBreadcrumb.php:168]: a dockblock could help

File: src/Location.php [hunk 4]
```diff
@@ -399,6 +399,19 @@ public static function displayTabContentForItem(CommonGLPI $item, $tabnum = 1, $
...
     }
 
+    /**
+     * get item location
+     *
+     * @param CommonDBTM  $item
+     *
+     * @return CommonDBTM
+     **/
+    public static function getItemLocation(CommonDBTM $item)
```
[the remark this answers is not in the document]
Comment (stonebuzz) [reply src/Location.php]: done

File: src/Location.php [hunk 5]
```diff
@@ -399,6 +399,23 @@ public static function displayTabContentForItem(CommonGLPI $item, $tabnum = 1, $
         return true;
     }
 
+    /**
+     * get item location
+     *
+     * @param CommonDBTM  $item
+     *
+     * @return Location if succeed else false
```
Comment (orthagh) [inline src/Location.php:407]:
```suggestion
     * @return Location|false
```

**Review by orthagh — APPROVED** [review]

**Comment by trasher** [comment]
Please fix tests ;)

**Comment by stonebuzz** [comment]
Ex (from Computer side ) with small location name

image

With long location name

image

File: templates/layout/parts/dcbreadcrumbs.html.twig [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ #}
+<div class="row">
+   {% for breadcrumb_item in breadcrumbs %}
+
+      <div class="col-auto p-1">
+         {{ breadcrumb_item['link']|raw }}
+
+         {% if breadcrumb_item['location']['name'] is defined %}
+            <br>
+            <span class='p-0 float-left badge bg-blue-lt d-inline-block text-truncate' data-toggle='tooltip' title='{{ breadcrumb_item['location']['name'] }}' style='max-width: 100px;'>
```
Comment (orthagh) [inline templates/layout/parts/dcbreadcrumbs.html.twig:41]:
```suggestion
            <span class='p-0 float-left badge bg-blue-lt d-inline-block text-truncate' data-bs-toggle='tooltip' title='{{ breadcrumb_item['location']['name'] }}' style='max-width: 100px;'>
```

**Review by orthagh — APPROVED** [review]

File: src/Features/DCBreadcrumb.php [hunk 7]
```diff
@@ -111,11 +141,85 @@ public function getDcBreadcrumb()
...
+     * Clean Location information foreach item breadcrumb
+     * If Current item Location = Parent item Location              -> do not display Location
+     * If Current item Location is child of Parent item Location    -> display only Location name instead of completename
+     * Else Current item Location != Parent item Location           -> display Location completename
+     *
+     * @param array $breadcrumb
+     *
+     * @return array cleanded $breadcrumb
+     */
+    public function cleanLocationInformations(&$breadcrumb)
```
Comment (cedric-anne) [inline src/Features/DCBreadcrumb.php:177]: 1. Your return the updated array, so no need to use a reference.
2. Method should be private as it is used only here.
```suggestion
    private function cleanLocationInformations(array $breadcrumb)
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]
Functionnaly, it seems OK, but technically, it is not OK for me. Indeed, you changed existing public methods signature, and, even if they does not seems to be used in plugins, it is preferable to avoid such changes.

I will adapt the code and fix tests.

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Pull request merged**